File: packages/queries/tasks.ts
import {
  type CreateTaskRequest,
  type Task,
  TaskStatus,
  type TaskWithChats,
  type UpdateTaskRequest,
} from '@openflow/generated';
import { createLogger } from '@openflow/utils';
import { createTaskSchema, updateTaskSchema } from '@openflow/validation';
import { invoke } from './utils.js';

/**
 * Logger for task query operations.
 * Logs at DEBUG level for query calls, INFO for successes, ERROR for failures.
 */
const logger = createLogger('queries:tasks');

/**
 * Task query wrappers for Tauri IPC.
 * Thin wrappers around invoke() calls with type safety.
 * Input validation is performed using Zod schemas before invoking Tauri commands.
 *
 * All functions include:
 * - Try/catch error handling with re-throw for React Query
 * - Logging at appropriate levels (DEBUG on call, INFO on success, ERROR on failure)
 * - Input validation via Zod schemas for mutations
 *
 * @example
 * ```ts
 * // List tasks for a project
 * const tasks = await taskQueries.list('project-123');
 *
 * // List tasks with status filter
 * const inProgressTasks = await taskQueries.list('project-123', 'in_progress');
 *
 * // Get a single task with its chats
 * const taskWithChats = await taskQueries.get('task-123');
 *
 * // Create a new task
 * const newTask = await taskQueries.create({
 *   projectId: 'project-123',
 *   title: 'My Task',
 * });
 *
 * // Update a task
 * const updated = await taskQueries.update('task-123', {
 *   title: 'Updated Title',
 *   status: 'in_progress',
 * });
 *
 * // Archive a task
 * const archived = await taskQueries.archive('task-123');
 *
 * // Duplicate a task
 * const duplicated = await taskQueries.duplicate('task-123');
 * ```
 */
export const taskQueries = {
  /**
   * List tasks for a project with optional filtering.
   * @param projectId - Project ID to filter tasks
   * @param status - Optional status filter (todo, in_progress, completed, blocked)
   * @param includeArchived - Whether to include archived tasks (default: false)
   * @returns Promise resolving to array of tasks
   * @throws Error if the query fails (re-thrown for React Query)
   */
  list: async (
    projectId: string,
    status?: TaskStatus,
    includeArchived?: boolean
  ): Promise<Task[]> => {
    logger.debug('Listing tasks', {
      projectId,
      status: status ?? 'all',
      includeArchived: includeArchived ?? false,
    });

    try {
      const tasks = await invoke<Task[]>('list_tasks', { projectId, status, includeArchived });

      logger.info('Tasks listed successfully', {
        projectId,
        count: tasks.length,
        titles: tasks.slice(0, 3).map((t) => t.title),
        hasMore: tasks.length > 3,
        byStatus: {
          todo: tasks.filter((t) => t.status === TaskStatus.Todo).length,
          inprogress: tasks.filter((t) => t.status === TaskStatus.Inprogress).length,
          inreview: tasks.filter((t) => t.status === TaskStatus.Inreview).length,
          done: tasks.filter((t) => t.status === TaskStatus.Done).length,
          cancelled: tasks.filter((t) => t.status === TaskStatus.Cancelled).length,
        },
      });

      return tasks;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to list tasks', {
        projectId,
        status,
        includeArchived,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Get a single task by ID with its associated chats.
   * @param id - Task ID
   * @returns Promise resolving to the task with its chats
   * @throws Error if the query fails (re-thrown for React Query)
   */
  get: async (id: string): Promise<TaskWithChats> => {
    logger.debug('Getting task', { id });

    try {
      const taskWithChats = await invoke<TaskWithChats>('get_task', { id });

      logger.info('Task retrieved successfully', {
        id: taskWithChats.task.id,
        title: taskWithChats.task.title,
        status: taskWithChats.task.status,
        chatCount: taskWithChats.chats?.length ?? 0,
        projectId: taskWithChats.task.projectId,
      });

      return taskWithChats;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to get task', {
        id,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Create a new task.
   * Input is validated against createTaskSchema before invoking.
   * @param request - Task creation request with projectId and title
   * @returns Promise resolving to the created task
   * @throws Error if validation or query fails (re-thrown for React Query)
   */
  create: async (request: CreateTaskRequest): Promise<Task> => {
    logger.debug('Creating task', {
      projectId: request.projectId,
      title: request.title,
      hasWorkflowId: 'workflowId' in request && !!request.workflowId,
    });

    try {
      const validated = createTaskSchema.parse(request);
      const task = await invoke<Task>('create_task', { request: validated });

      logger.info('Task created successfully', {
        id: task.id,
        title: task.title,
        projectId: task.projectId,
        status: task.status,
      });

      return task;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to create task', {
        projectId: request.projectId,
        title: request.title,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Update an existing task.
   * Input is validated against updateTaskSchema before invoking.
   * @param id - Task ID to update
   * @param request - Update request with fields to change
   * @returns Promise resolving to the updated task
   * @throws Error if validation or query fails (re-thrown for React Query)
   */
  update: async (id: string, request: UpdateTaskRequest): Promise<Task> => {
    logger.debug('Updating task', {
      id,
      hasTitle: 'title' in request,
      hasStatus: 'status' in request,
      hasDescription: 'description' in request,
    });

    try {
      const validated = updateTaskSchema.parse(request);
      const task = await invoke<Task>('update_task', { id, request: validated });

      logger.info('Task updated successfully', {
        id: task.id,
        title: task.title,
        status: task.status,
      });

      return task;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to update task', {
        id,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Archive a task (soft delete).
   * Archived tasks are hidden from list queries by default.
   * @param id - Task ID to archive
   * @returns Promise resolving to the archived task
   * @throws Error if the query fails (re-thrown for React Query)
   */
  archive: async (id: string): Promise<Task> => {
    logger.debug('Archiving task', { id });

    try {
      const task = await invoke<Task>('archive_task', { id });

      logger.info('Task archived successfully', {
        id: task.id,
        title: task.title,
      });

      return task;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to archive task', {
        id,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Unarchive a task (restore from archive).
   * Makes the task visible in list queries again.
   * @param id - Task ID to unarchive
   * @returns Promise resolving to the unarchived task
   * @throws Error if the query fails (re-thrown for React Query)
   */
  unarchive: async (id: string): Promise<Task> => {
    logger.debug('Unarchiving task', { id });

    try {
      const task = await invoke<Task>('unarchive_task', { id });

      logger.info('Task unarchived successfully', {
        id: task.id,
        title: task.title,
      });

      return task;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to unarchive task', {
        id,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Permanently delete a task.
   * This will cascade delete all associated chats and messages.
   * @param id - Task ID to delete
   * @returns Promise resolving when deletion is complete
   * @throws Error if the query fails (re-thrown for React Query)
   */
  delete: async (id: string): Promise<void> => {
    logger.debug('Deleting task', { id });

    try {
      await invoke<void>('delete_task', { id });

      logger.info('Task deleted successfully', { id });
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to delete task', {
        id,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Duplicate an existing task.
   * Creates a copy with "(copy)" appended to the title and status reset to "todo".
   * @param id - Task ID to duplicate
   * @returns Promise resolving to the duplicated task
   * @throws Error if the query fails (re-thrown for React Query)
   */
  duplicate: async (id: string): Promise<Task> => {
    logger.debug('Duplicating task', { id });

    try {
      const task = await invoke<Task>('duplicate_task', { id });

      logger.info('Task duplicated successfully', {
        originalId: id,
        newId: task.id,
        title: task.title,
        projectId: task.projectId,
      });

      return task;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to duplicate task', {
        id,
        error: errorMessage,
      });
      throw error;
    }
  },
};

File: src-tauri/src/commands/tasks.rs
//! Tauri commands for task operations.
//!
//! These commands provide the IPC interface for task CRUD operations
//! and task execution control.
//!
//! Each command is a thin wrapper around openflow_core::services::task
//! and task_executor functions.

use tauri::State;

use crate::commands::AppState;
use openflow_contracts::{
    AgentEventRecord, CreateStepRequest, CreateTaskRequest, Permission, Task, TaskStatus, TaskStep,
    TaskWithChats, TaskWithSteps, UpdateTaskRequest,
};
use openflow_core::services::{agent_session, task};

/// List tasks for a project with optional filters.
///
/// # Arguments
/// * `project_id` - The project to list tasks for
/// * `status` - Optional status filter
/// * `include_archived` - Whether to include archived tasks (default: false)
#[tauri::command]
pub async fn list_tasks(
    state: State<'_, AppState>,
    project_id: String,
    status: Option<TaskStatus>,
    include_archived: Option<bool>,
) -> Result<Vec<Task>, String> {
    let pool = state.db.lock().await;
    task::list(
        &pool,
        &project_id,
        status,
        include_archived.unwrap_or(false),
    )
    .await
    .map_err(|e| e.to_string())
}

/// Get a single task by ID with its associated chats.
///
/// Returns the task with all its chats if found, or an error if not found.
#[tauri::command]
pub async fn get_task(state: State<'_, AppState>, id: String) -> Result<TaskWithChats, String> {
    let pool = state.db.lock().await;
    task::get(&pool, &id).await.map_err(|e| e.to_string())
}

/// Create a new task.
///
/// Returns the newly created task with generated ID and timestamps.
#[tauri::command]
pub async fn create_task(
    state: State<'_, AppState>,
    request: CreateTaskRequest,
) -> Result<Task, String> {
    let pool = state.db.lock().await;
    task::create(&pool, request)
        .await
        .map_err(|e| e.to_string())
}

/// Update an existing task.
///
/// Only the provided fields will be updated. Returns the updated task.
#[tauri::command]
pub async fn update_task(
    state: State<'_, AppState>,
    id: String,
    request: UpdateTaskRequest,
) -> Result<Task, String> {
    let pool = state.db.lock().await;
    task::update(&pool, &id, request)
        .await
        .map_err(|e| e.to_string())
}

/// Archive a task by ID.
///
/// Sets the archived_at timestamp. Archived tasks are excluded from default listings.
#[tauri::command]
pub async fn archive_task(state: State<'_, AppState>, id: String) -> Result<Task, String> {
    let pool = state.db.lock().await;
    task::archive(&pool, &id).await.map_err(|e| e.to_string())
}

/// Unarchive a task by ID.
///
/// Clears the archived_at timestamp, restoring the task to active listings.
#[tauri::command]
pub async fn unarchive_task(state: State<'_, AppState>, id: String) -> Result<Task, String> {
    let pool = state.db.lock().await;
    task::unarchive(&pool, &id).await.map_err(|e| e.to_string())
}

/// Delete a task by ID.
///
/// This will cascade delete all associated chats and messages.
#[tauri::command]
pub async fn delete_task(state: State<'_, AppState>, id: String) -> Result<(), String> {
    let pool = state.db.lock().await;
    task::delete(&pool, &id).await.map_err(|e| e.to_string())
}

/// Duplicate a task by ID.
///
/// Creates a copy of the task with:
/// - New unique ID
/// - Title appended with "(copy)"
/// - Status reset to "pending"
/// - Associated chats are NOT duplicated (new task starts fresh)
///
/// Returns the newly created duplicate task.
#[tauri::command]
pub async fn duplicate_task(state: State<'_, AppState>, id: String) -> Result<Task, String> {
    let pool = state.db.lock().await;
    task::duplicate(&pool, &id).await.map_err(|e| e.to_string())
}

// =============================================================================
// Task Execution Commands
// =============================================================================

/// Get a task with all its steps.
///
/// Returns the task along with all associated steps ordered by step_index.
/// This is the primary query for the task execution UI.
#[tauri::command]
pub async fn get_task_with_steps(
    state: State<'_, AppState>,
    id: String,
) -> Result<TaskWithSteps, String> {
    let pool = state.db.lock().await;
    task::get_with_steps(&pool, &id)
        .await
        .map_err(|e| e.to_string())
}

/// Start executing a task.
///
/// This command:
/// 1. Validates the task can be started (pending or paused status)
/// 2. Updates task status to running
/// 3. Spawns background execution
///
/// Returns the updated task after status change.
/// Execution continues in background - use events to monitor progress.
#[tauri::command]
pub async fn start_task(state: State<'_, AppState>, id: String) -> Result<Task, String> {
    state
        .task_executor
        .start_task(&id)
        .await
        .map_err(|e| e.to_string())
}

/// Pause a running task.
///
/// This command:
/// 1. Kills the current agent if running
/// 2. Marks the current step as failed (interrupted)
/// 3. Updates task status to paused
///
/// The task can be resumed later with resume_task.
#[tauri::command]
pub async fn pause_task(state: State<'_, AppState>, id: String) -> Result<Task, String> {
    state
        .task_executor
        .pause_task(&id)
        .await
        .map_err(|e| e.to_string())
}

/// Resume a paused task.
///
/// Continues execution from the current step. If the current step was
/// marked as failed during pause, it will be reset to pending and re-executed.
#[tauri::command]
pub async fn resume_task(state: State<'_, AppState>, id: String) -> Result<Task, String> {
    state
        .task_executor
        .resume_task(&id)
        .await
        .map_err(|e| e.to_string())
}

/// Cancel a task.
///
/// This command:
/// 1. Kills the current agent if running
/// 2. Marks remaining pending steps as skipped
/// 3. Updates task status to cancelled (terminal - cannot be resumed)
#[tauri::command]
pub async fn cancel_task(state: State<'_, AppState>, id: String) -> Result<Task, String> {
    state
        .task_executor
        .cancel_task(&id)
        .await
        .map_err(|e| e.to_string())
}

/// Check if a task is currently running.
#[tauri::command]
pub async fn is_task_running(state: State<'_, AppState>, id: String) -> Result<bool, String> {
    Ok(state.task_executor.is_running(&id).await)
}

/// Get the count of currently running tasks.
#[tauri::command]
pub async fn running_task_count(state: State<'_, AppState>) -> Result<usize, String> {
    Ok(state.task_executor.running_count().await)
}

/// List all currently running task IDs.
#[tauri::command]
pub async fn list_running_tasks(state: State<'_, AppState>) -> Result<Vec<String>, String> {
    Ok(state.task_executor.list_running().await)
}

// =============================================================================
// Task Step Commands
// =============================================================================

/// Create a step for a task.
///
/// Steps define the execution sequence for a task. Each step has a prompt
/// that will be sent to the configured provider (e.g., Claude Code).
#[tauri::command]
pub async fn create_task_step(
    state: State<'_, AppState>,
    task_id: String,
    request: CreateStepRequest,
) -> Result<TaskStep, String> {
    let pool = state.db.lock().await;
    task::create_step_from_request(&pool, &task_id, request)
        .await
        .map_err(|e| e.to_string())
}

/// Get a specific step by ID.
#[tauri::command]
pub async fn get_task_step(state: State<'_, AppState>, id: String) -> Result<TaskStep, String> {
    let pool = state.db.lock().await;
    task::get_step(&pool, &id).await.map_err(|e| e.to_string())
}

/// List all steps for a task.
///
/// Returns steps ordered by step_index (ascending).
#[tauri::command]
pub async fn list_task_steps(
    state: State<'_, AppState>,
    task_id: String,
) -> Result<Vec<TaskStep>, String> {
    let pool = state.db.lock().await;
    task::list_steps(&pool, &task_id)
        .await
        .map_err(|e| e.to_string())
}

/// Delete a specific step.
#[tauri::command]
pub async fn delete_task_step(state: State<'_, AppState>, id: String) -> Result<(), String> {
    let pool = state.db.lock().await;
    task::delete_step(&pool, &id)
        .await
        .map_err(|e| e.to_string())
}

/// Delete all steps for a task.
///
/// Returns the number of steps deleted.
#[tauri::command]
pub async fn delete_all_task_steps(
    state: State<'_, AppState>,
    task_id: String,
) -> Result<i32, String> {
    let pool = state.db.lock().await;
    task::delete_all_steps(&pool, &task_id)
        .await
        .map_err(|e| e.to_string())
}

// =============================================================================
// Task Event Commands
// =============================================================================

/// Get agent events for a task step.
///
/// Returns events for the session linked to the step at the given index.
/// Use after_sequence to get only new events since the last fetch.
///
/// # Arguments
/// * `task_id` - The task ID
/// * `step_index` - The step index (0-based)
/// * `after_sequence` - Only return events with sequence > this value
#[tauri::command]
pub async fn get_task_step_events(
    state: State<'_, AppState>,
    task_id: String,
    step_index: i32,
    after_sequence: Option<i32>,
) -> Result<Vec<AgentEventRecord>, String> {
    let pool = state.db.lock().await;

    // Get the task with steps
    let task_with_steps = task::get_with_steps(&pool, &task_id)
        .await
        .map_err(|e| e.to_string())?;

    // Find the step at the given index
    let step = task_with_steps
        .steps
        .iter()
        .find(|s| s.step_index == step_index)
        .ok_or_else(|| format!("Step at index {} not found", step_index))?;

    // If step has no session, return empty vec
    let session_id = match &step.session_id {
        Some(id) => id,
        None => return Ok(Vec::new()),
    };

    // Get events for the session
    agent_session::get_events(&pool, session_id, after_sequence.map(|s| s as i64))
        .await
        .map_err(|e| e.to_string())
}

// =============================================================================
// Permission Commands
// =============================================================================

/// Respond to a permission request for a running task.
///
/// When an agent requests permission (e.g., to write a file), this command
/// approves or denies the request and sends the response to the agent.
///
/// # Arguments
/// * `task_id` - The task ID
/// * `permission_id` - The permission request ID
/// * `approved` - Whether to approve (true) or deny (false) the request
#[tauri::command]
pub async fn respond_to_task_permission(
    state: State<'_, AppState>,
    task_id: String,
    permission_id: String,
    approved: bool,
) -> Result<Permission, String> {
    state
        .task_executor
        .respond_to_permission(&task_id, &permission_id, approved)
        .await
        .map_err(|e| e.to_string())
}

/// Get the pending permission for a running task.
///
/// Returns the current pending permission request if one exists, or None
/// if there is no pending permission.
#[tauri::command]
pub async fn get_task_pending_permission(
    state: State<'_, AppState>,
    task_id: String,
) -> Result<Option<Permission>, String> {
    state
        .task_executor
        .get_pending_permission(&task_id)
        .await
        .map_err(|e| e.to_string())
}
